cmd/xl-v1-healing-common.go

```diff
@@ -252,39 +252,39 @@ func xlHealStat(xl xlObjects, partsMetadata []xlMetaV1, errs []error) HealObject
 // calculating blake2b checksum.
 func disksWithAllParts(onlineDisks []StorageAPI, partsMetadata []xlMetaV1, errs []error, bucket, object string) ([]StorageAPI, []error, error) {
 	availableDisks := make([]StorageAPI, len(onlineDisks))
-	for index, onlineDisk := range onlineDisks {
+	for diskIndex, onlineDisk := range onlineDisks {
 		if onlineDisk == nil {
 			continue
 		}
 		// disk has a valid xl.json but may not have all the
 		// parts. This is considered an outdated disk, since
 		// it needs healing too.
-		for _, part := range partsMetadata[index].Parts {
+		for partIndex, part := range partsMetadata[diskIndex].Parts {
 			// compute blake2b sum of part.
 			partPath := filepath.Join(object, part.Name)
-			hash := newHash(partsMetadata[index].Erasure.Algorithm)
+			hash := newHash(partsMetadata[diskIndex].Erasure.Checksum[partIndex].Algorithm)
```

Comment:
For example, if I enforce mc to upload parts with a reverse order, I get this in xl.json:

```
{
  "version": "1.0.1",
  "format": "xl",
  "stat": {
    "size": 134217728,
    "modTime": "2017-06-10T11:23:46.742721655Z"
  },
  "erasure": {
    "algorithm": "klauspost/reedsolomon/vandermonde",
    "data": 2,
    "parity": 2,
    "blockSize": 10485760,
    "index": 1,
    "distribution": [
      1,
      2,
      3,
      4
    ],
    "checksum": [
      {
        "name": "part.2",
        "algorithm": "blake2b",
        "hash": "76a510f220d2ce6423a92b2df2b0a3fe5ce0c77a575a4ead76dca3e80bcc56c79e5a245c7ca0d7add2e73aebf43abf6e9a009d916799fba1922fe6183498356b"
      },
      {
        "name": "part.1",
        "algorithm": "blake2b",
        "hash": "dfec7a954c6c9c616d586751960682dab9cd2e825f6d3aab97aec523f50c092245d05600762235790b45919531b4516bb56db207f5afb2ec7395254f532f4b8a"
      }
    ]
  },
  "minio": {
    "release": "DEVELOPMENT.GOGET"
  },
  "meta": {
    "content-type": "application/octet-stream",
    "etag": "51b1206b454b32ac37305a2b544b0511-2"
  },
  "parts": [
    {
      "number": 1,
      "name": "part.1",
      "etag": "48255d8705e103628f6b443fb70cb533",
      "size": 67108864
    },
    {
      "number": 2,
      "name": "part.2",
      "etag": "8e018c99de4371f5f7634f74aff04683",
      "size": 67108864
    }
  ]
}
```